From: Jan Grulich <jgrulich@redhat.com>
Date: Thu, 20 Aug 2026 18:33:48 +0000
Subject: Bug 2054624 - WebRTC backport: Video capture: cap resolution and
 null-check header metadata a=RyanVM DONTBUILD

Reject video dimensions that could overflow frame size calculations.
Handle missing SPA_META_Header which is optional.

This is a simple backport of an WebRTC upstream change.

Upstream commit: 01455efea7797cff42c5228b14c85ab3cdb4bdac

Original Revision: https://phabricator.services.mozilla.com/D319163

Differential Revision: https://phabricator.services.mozilla.com/D319394
Mercurial Revision: https://hg.mozilla.org/mozilla-central/rev/09b103344aad
---
 .../linux/video_capture_pipewire.cc              | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/modules/video_capture/linux/video_capture_pipewire.cc b/modules/video_capture/linux/video_capture_pipewire.cc
index 8118f5e2b2..dfacadfc27 100644
--- a/modules/video_capture/linux/video_capture_pipewire.cc
+++ b/modules/video_capture/linux/video_capture_pipewire.cc
@@ -52,6 +52,10 @@
 namespace webrtc {
 namespace videocapturemodule {
 
+// A reasonable maximum size so "width * height * kBytesPerPixel" doesn't
+// overflow
+constexpr int32_t kMaxVideoCaptureDimension = 16384;
+
 struct {
   uint32_t spa_format;
   VideoType video_type;
@@ -351,6 +355,14 @@ void VideoCaptureModulePipeWire::OnFormatChanged(const struct spa_pod* format) {
     return;
   }
 
+  if (configured_capability_.width <= 0 || configured_capability_.height <= 0 ||
+      configured_capability_.width > kMaxVideoCaptureDimension ||
+      configured_capability_.height > kMaxVideoCaptureDimension) {
+    RTC_LOG(LS_ERROR) << "Unsupported video resolution.";
+    configured_capability_.videoType = VideoType::kUnknown;
+    return;
+  }
+
   RTC_LOG(LS_VERBOSE) << "Configured capture format = "
                       << static_cast<int>(configured_capability_.videoType);
 
@@ -482,8 +494,8 @@ void VideoCaptureModulePipeWire::ProcessBuffers() {
       SetApplyRotation(rotation != kVideoRotation_0);
     }
 
-    if (h->flags & SPA_META_HEADER_FLAG_CORRUPTED) {
-      RTC_LOG(LS_INFO) << "Dropping corruped frame.";
+    if (h && (h->flags & SPA_META_HEADER_FLAG_CORRUPTED)) {
+      RTC_LOG(LS_INFO) << "Dropping corrupted frame.";
       pw_stream_queue_buffer(stream_, buffer);
       continue;
     }
